upup/pkg/fi/cloudup/do/cloud.go

```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
+	}
+
+	_, _, err = c.Client.DropletActions.Shutdown(context.TODO(), dropletID)
+	if err != nil {
+		return fmt.Errorf("error stopping instance %q: %v", dropletID, err)
+	}
+
+	// Wait for 5 min to stop the instance
+	for i := 0; i < 5; i++ {
+		droplet, _, err := c.Client.Droplets.Get(context.TODO(), dropletID)
+		if err != nil {
+			return fmt.Errorf("error describing instance %q: %v", dropletID, err)
+		}
+
+		klog.V(8).Infof("stopping DO instance %q, current Status: %q", droplet, droplet.Status)
+
+		if droplet.Status == "off" {
+			break
+		}
+
+		if i == 5 {
+			return fmt.Errorf("fail to stop DO instance %v in 5 mins", dropletID)
+		}
+
+		time.Sleep(time.Minute * 1)
+	}
+
+	_, err = c.Client.Droplets.Delete(context.TODO(), dropletID)
+	if err != nil {
+		return fmt.Errorf("error stopping instance %q: %v", dropletID, err)
+	}
+
+	klog.V(8).Infof("deleted droplet instance %q", dropletID)
+
+	return nil
+}
+
+// DetachInstance is not implemented yet. It needs to cause a cloud instance to no longer be counted against the group's size limits.
+func (c *doCloudImplementation) DetachInstance(i *cloudinstances.CloudInstance) error {
+	klog.V(8).Info("digitalocean cloud provider DetachInstance not implemented yet")
+	return fmt.Errorf("digital ocean cloud provider does not support surging")
+}
+
+// ProviderID returns the kops api identifier for DigitalOcean cloud provider
+func (c *doCloudImplementation) ProviderID() kops.CloudProviderID {
+	return kops.CloudProviderDO
+}
+
+// Region returns the DO region we will target
+func (c *doCloudImplementation) Region() string {
+	return c.region
+}
+
+func (c *doCloudImplementation) DNS() (dnsprovider.Interface, error) {
+	provider := dns.NewProvider(c.Client)
+	return provider, nil
+}
+
+// Volumes returns an implementation of godo.StorageService
+func (c *doCloudImplementation) VolumeService() godo.StorageService {
+	return c.Client.Storage
+}
+
+// VolumeActions returns an implementation of godo.StorageActionsService
+func (c *doCloudImplementation) VolumeActionService() godo.StorageActionsService {
+	return c.Client.StorageActions
+}
+
+// DropletsService returns the droplets client interface.
+func (c *doCloudImplementation) DropletsService() godo.DropletsService {
+	return c.Client.Droplets
+}
+
+func (c *doCloudImplementation) DropletActionService() godo.DropletActionsService {
+	return c.Client.DropletActions
+}
+
+func (c *doCloudImplementation) LoadBalancersService() godo.LoadBalancersService {
+	return c.Client.LoadBalancers
+}
+
+func (c *doCloudImplementation) DomainService() godo.DomainsService {
+	return c.Client.Domains
+}
+
+func (c *doCloudImplementation) ActionsService() godo.ActionsService {
+	return c.Client.Actions
+}
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *doCloudImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("Not implemented")
```

Comment:
Go errors should always start with lowercase letters

```suggestion
	return nil, errors.New("not implemented")
```

Reply:
sure, updated.